presto-main/src/main/java/com/facebook/presto/operator/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,751 @@
+    {
+        return pagePartitioner.getInfo();
+    }
+
+    @Override
+    public void finish()
+    {
+        finished = true;
+        pagePartitioner.flush();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finished && isBlocked().isDone();
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        ListenableFuture<?> blocked = pagePartitioner.isFull();
+        return blocked.isDone() ? NOT_BLOCKED : blocked;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished && isBlocked().isDone();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+
+        if (page.getPositionCount() == 0) {
+            return;
+        }
+
+        page = pagePreprocessor.apply(page);
+        pagePartitioner.partitionPage(page);
+
+        // TODO: PartitionedOutputOperator reports incorrect output data size #11770
+        operatorContext.recordOutput(page.getSizeInBytes(), page.getPositionCount());
+
+        systemMemoryContext.setBytes(pagePartitioner.getRetainedSizeInBytes());
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        return null;
+    }
+
+    private static class PartitionBuffer
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(PartitionBuffer.class).instanceSize();
+
+        private final int partition;
+        private final AtomicLong rowsAdded;
+        private final AtomicLong pagesAdded;
+        private final PagesSerde serde;
+        private final Lifespan lifespan;
+        private final int capacity;
+        private final int channelCount;
+
+        private int[] positions;   // the default positions array for top level BlockEncodingBuffers
+        private int[] serializedRowSizeInBytes;  // The sizes of the rows if they were serialized
+        private int positionCount;  // number of positions to be copied for this partition
+        private BlockEncodingBuffers[] blockEncodingBuffers;
+
+        private int bufferedRowCount;
+        private boolean bufferFull;
+
+        PartitionBuffer(int partition, int channelCount, int capacity, AtomicLong pagesAdded, AtomicLong rowsAdded, PagesSerde serde, Lifespan lifespan)
+        {
+            this.partition = partition;
+            this.channelCount = channelCount;
+            this.capacity = capacity;
+            this.pagesAdded = requireNonNull(pagesAdded, "pagesAdded is null");
+            this.rowsAdded = requireNonNull(rowsAdded, "rowsAdded is null");
+            this.serde = requireNonNull(serde, "serde is null");
+            this.lifespan = requireNonNull(lifespan, "lifespan is null");
+        }
+
+        public void resetPositions(int positionCount)
+        {
+            positions = ensureCapacity(positions, positionCount);
+            this.positionCount = 0;
+        }
+
+        public void appendPosition(int position)
```

Comment:
Maybe add comments to `appendPosition` and `appendData` methods explain they are used in in this two-phased "partition-and-serialize" algorithm. `appendPosition` is only for register the positions going to this partition, while ` appendData` really copy the data.

Reply:
How about renaming appendPosition() to addPosition()? Also rename BlockEncodingBuffers.copyValuesInBatch() to appendDataInBatch to make them agree with each other? 
In partitionPage() there is already a comment saying
`// Populate positions to copy for each destination partition.`